wurst/file/Base64.wurst

```diff
@@ -0,0 +1,270 @@
+    "w","x","y","z","0","1","2","3","4","5","6","7","8","9","+","/"
+]
+
+// RFC 4648 compliant Base64 reverse charmap.
+constant REVERSE_CHARMAP = [
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 62, 0, 0, 0, 63, 52, 53, 54,
+    55, 56, 57, 58, 59, 60, 61, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2,
+    3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
+    20, 21, 22, 23, 24, 25, 0, 0, 0, 0, 0, 0, 26, 27, 28, 29, 30,
+    31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
+    48, 49, 50, 51, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+]
+
+public class Base64Encoder
+    private ChunkedString chunkedString
+
+    // We append only one char each time, and strings in WC3 get fully copied on each concatenation.
+    // To optimize it and lower the times of copying quite big chunks, append by smaller chunks.
+    private var stringBuffer = ""
+    private var chars = 0
+    private static constant MAX_CHARS = 32
+
+    private var buffer = 0
+    private var bytes = 0
+    private static constant MAX_BYTES = 3
+
+    construct()
+        this.chunkedString = new ChunkedString()
+
+    construct(int maxChunkLength)
+        this.chunkedString = new ChunkedString(maxChunkLength)
+
+    ondestroy
+        if chunkedString != null // The string is null after calling `intoData()`.
+            destroy chunkedString
+
+    // It's way too much overhead in dispatching two lambdas per each byte, 'inlining' them to deal with this.
+    private function writeData(ByteBuffer data)
+        execute() ->
+            var i = 0
+            while data.hasByte() and i < ENCODES_PER_ROUND
+                writeByteUnsafe(data.readByte())
+                i++
+
+            if data.hasByte()
+                writeData(data)
+
+    private function flushStringBuffer()
+        chunkedString.append(stringBuffer)
+        stringBuffer = ""
+        chars = 0
+
+    private function append(string char)
+        stringBuffer += char
+        chars++
+        if chars == MAX_CHARS
+            flushStringBuffer()
+
+    private function encode(int byte, int count)
+        var remaining = byte
+
+        for i = 0 to count
+            let c = remaining.bitAnd(compiletime("11111100 00000000 00000000".fromBitString())).shiftr(18)
+            append(CHARMAP[c])
+            remaining = remaining.shiftl(6)
+
+    /** Writes an unsigned byte to be serialized into Base64. 
+    You must be sure that the provided integer is in the range [0, 255]. */
```

Comment:
> You must be sure that the provided integer

Ideally find a way to add checks that can be optimised away

Reply:
They're private here. ByteBuffer have similar public functions and I've added checks there using SafetyChecks package.